src/netlink.c

```diff
@@ -0,0 +1,365 @@
+\*-------------------------------------------------------------------------*/
+static int meth_send(lua_State *L) {
+    p_netlink nl = (p_netlink)auxiliar_checkclass(L, "netlink{connected}", 1);
+    size_t payload_size;
+    const char *payload = luaL_checklstring(L, 2, &payload_size);
+    int flags = luaL_optinteger(L, 3, 0);
+    struct nlmsghdr nlh[NLMSG_SPACE(payload_size)];
+    p_timeout tm = &nl->tm; 
+    size_t sent = 0;
+    int err;
+
+    memset(nlh, 0, NLMSG_SPACE(payload_size));
+    nlh->nlmsg_len = NLMSG_SPACE(payload_size);
+    nlh->nlmsg_pid = nl->srcpid;
+    nlh->nlmsg_flags = flags;
+    memcpy(NLMSG_DATA(nlh), payload, payload_size);
+    timeout_markstart(tm);
+
+    err = socket_send(&nl->fd, (char *)nlh, nlh->nlmsg_len, &sent, tm);
+    if (err != IO_DONE) {
+        lua_pushnil(L);
+        lua_pushliteral(L, "error sending message");
+        return 2;
+    }
+
+    lua_pushinteger(L, sent);
+    return 1;
+}
+
+/*-------------------------------------------------------------------------*\
+* Send data through unconnected netlink socket
+\*-------------------------------------------------------------------------*/
+static int meth_sendto(lua_State *L) {
+    p_netlink nl = (p_netlink)auxiliar_checkclass(L, "netlink{unconnected}", 1);
+    size_t payload_size;
+    const char *payload = luaL_checklstring(L, 2, &payload_size);
+    int dstpid = luaL_checkinteger(L, 3);
+    int groups = luaL_optinteger(L, 4, 0);
+    int flags = luaL_optinteger(L, 5, 0);
+    struct nlmsghdr nlh[NLMSG_SPACE(payload_size)];
+    struct sockaddr_nl addr;
+    p_timeout tm = &nl->tm; 
+    size_t sent = 0;
+    int err;
+
+    memset(&addr, 0, sizeof(addr));
+    addr.nl_pid = dstpid;
+    addr.nl_family = AF_NETLINK;
+    addr.nl_groups = groups;
+
+    memset(nlh, 0, NLMSG_SPACE(payload_size));
+    nlh->nlmsg_len = NLMSG_SPACE(payload_size);
+    nlh->nlmsg_pid = nl->srcpid;
+    nlh->nlmsg_flags = flags;
+    memcpy(NLMSG_DATA(nlh), payload, payload_size);
+    timeout_markstart(tm);
+
+    err = socket_sendto(&nl->fd, (char *)nlh, nlh->nlmsg_len, &sent, 
+            (SA *)&addr, sizeof(addr), tm);
+    if (err != IO_DONE) {
+        lua_pushnil(L);
+        lua_pushliteral(L, "error sending message");
+        return 2;
+    }
+
+    lua_pushinteger(L, sent);
+    return 1;
+}
+
+/*-------------------------------------------------------------------------*\
+* Receives data from a netlink socket
+\*-------------------------------------------------------------------------*/
+static int meth_receive(lua_State *L) {
+    p_netlink nl = (p_netlink)auxiliar_checkclass(L, "netlink{connected}", 1);
+    struct nlmsghdr nlh[NLMSG_SPACE(MAX_PAYLOAD)];
+    size_t got;
+    p_timeout tm = &nl->tm;
+    int err;
+
+    memset(nlh, 0, NLMSG_SPACE(MAX_PAYLOAD));
+    nlh->nlmsg_len = NLMSG_SPACE(MAX_PAYLOAD);
+    timeout_markstart(tm);
+    err = socket_recv(&nl->fd, (char *)nlh, nlh->nlmsg_len, &got, tm);
```

Comment:
How about multipart messages (`NLM_F_MULTI`), acknowlegment requests (`NLM_F_ACK`) and other flag options? They are all ignored and potentially discards data of the multipart messages?
Why ignore field `nlmsg_pid`? How about `nlmsg_type` and `nlmsg_seq` again?

Reply:
Those flags aren't treated here. The flags are read by the netlink core (inside recv, recvfrom and recvmsg), that's why they're specified  only when we send the message. Again, those fields are opaque to netlink core, and only carry any meaning if we arbitrarily add meaning to them.

Reply:
Again, would you care to enlighten me by point me from where did you get this about `recv`, `recvfrom`, and `recvmsg` processing those flags? Because from everything I've read about Netlink that's complete rubbish.

Reply:
It is in fact wrong, I actually misinterpreted a couple of lines from Kernel Korner's arcticle. Also, I believe that is done is kernel not in userspace. This is in 'man 7 netlink' for example:

> For reliable transfer the
       sender can request an acknowledgement from the receiver by setting
       the NLM_F_ACK flag.  An acknowledgment is an NLMSG_ERROR packet with
       the error field set to 0.  The application must generate acknowledge‐
       ments for received messages itself.  The kernel tries to send an
       NLMSG_ERROR message for every failed packet.  A user process should
       follow this convention too.